README.md
# ts3admin.class
The ts3admin.class is a powerful api for communication with Teamspeak 3 Servers from your website! Your creativity knows no bounds!

ts3admin.class.php written by Stefan 'par0noid' Zehnpfennig

You can use this software under the terms of the GNU General Public License v3.

Dev-Website: http://ts3admin.info

http://par0noid.info

## Installation via Composer

1.) Install composer requirements:

```
composer require par0noid/ts3admin
```

## Minimal script
```
<?php

//Include ts3admin composer library
require_once __DIR__ . '/vendor/autoload.php';
use par0noid\ts3admin\ts3admin;

//build a new ts3admin object
$tsAdmin = new ts3admin('localhost', 10011);

if($tsAdmin->getElement('success', $tsAdmin->connect())) {
    //login as serveradmin
    $tsAdmin->login('serveradmin', 'password');

    //select teamspeakserver
    $tsAdmin->selectServer(9987);

    //get clientlist
    $clients = $tsAdmin->clientList();

    echo count($clients['data']) . ' clients on selected server<br><br>';

    //print clients to browser
    foreach($clients['data'] as $client) {
        echo $client['client_nickname'] . '<br>';
    }
} else {
    echo 'Connection could not be established.';
}
```

CHANGELOG
Tested compatibility to version: 3.3.1 x64
__________________________________________

Version: 1.0.2.5 - Thu, 30 Aug 2018

~ changed description of logAdd, loglevels added (thanks to derrodirik)
~ changed description of banList (thanks to derrodirik)
~ changed description of clientGetNameFromUid (thanks to WrightProjects)
~ added filter function to serverGroupList (thanks to thomas-rolling)
~ overworked documentation, especially return-types (thanks to thomas-rolling)


Version: 1.0.2.4 - Tue, 05 Jun 2018

+ added customSet and customDelete function
+ added reason message to serverStop and serverProcessStop
~ fixed error in readChatMessage when you're in the default channel and want to read text from it
~ channelEdit comment embellished
- removed CLIENT_ICON_ID from clientdbinfo


Version: 1.0.2.3 - Thu, 02 Nov 2017

~ changed description of clientEdit (Thanks to Ibrahim for the hint)


Version: 1.0.2.2 - Thu, 02 Nov 2017

~ parameter seconds_empty changed to secondsempty in channelList (thanks to Mateusz)


Version: 1.0.2.1 - Tue, 07 Feb 2017

~ fixed composer.json


Version: 1.0.2.0 - Sun, 05 Feb 2017

+ added channelClientList function
+ added channelGroupAddClient as alias for setClientChannelGroup function


Version: 1.0.1.9 - Tue, 31 Jan 2017

+ added getIconByID function
+ added channelGroupGetIconByCGID function
+ added serverGroupGetIconBySGID function
+ added uploadIcon function
~ ftUploadFile seems to need no newline after sending fileTransferKey anymore, so I removed it
~ changed description of ftGetFileInfo
- removed channelGetIconByID function, replacement is getIconByID


Version: 1.0.1.8 - Fri, 04 Nov 2016

~ Changed misspelled property CHANNEL_CODEC (Thanks to Micky5991)
~ Force lowercase property-names for all create and edit functions (Thanks to Micky5991)


Version: 1.0.1.7 - Wed, 02 Nov 2016

+ Added constants for opus codec (thanks to Nik99)


Version: 1.0.1.6 - Wed, 02 Nov 2016

+ Added channelGetIconByChannelID
~ channelGetIconByID now accepts negative ID's


Version: 1.0.1.5 - Wed, 02 Nov 2016

+ Added channelGetIconByID


Version: 1.0.1.4 - Wed, 02 Nov 2016

+ Added readChatMessage function by toxiicdev
+ Added getQueryClid by toxiicdev
+ Added loadQueryData by toxiicdev
+ Added serverNotifyUnregister by toxiicdev


Version: 1.0.1.3 - Thu, 20 Oct 2016

~ fixed infinite loop if socket times out while sending a command (thanks to toxiicdev)


Version: 1.0.1.2 - Wed, 19 Oct 2016

+ Added some constants which can be found in the official ts3 documentation
~ changed documentation stuff


Version: 1.0.1.1 - Mon, 17 Oct 2016

~ changed documentation of channelCreate function
~ changed documentation of channelEdit function
~ changed documentation of serverCreate function
~ changed documentation of serverEdit function
~ changed documentation of clientEdit function
~ changed documentation of instanceEdit function


Version: 1.0.1.0 - Mon, 17 Oct 2016

+ Added subsystem parameter to bindingList
+ Added proto parameter to ftinitdownload
+ Added proto parameter to ftinitupload
~ Fixed bug with limit parameter in clientDbList (thanks to Nik99)


Version: 1.0.0.1 - Thu, 29 Sep 2016

~ Add a check to be sure the avatar exist before trying to init download (thanks to Viper61)


Version: 1.0.0.0 - Tue, 19 Jul 2016

+ added permReset function
+ added permGet function
+ added permFind function
+ added messageAdd function
+ added messageDelete function
+ added messageGet function
+ added messageUpdateFlag function
+ added messageList function
+ added servergroupautoaddperm function
+ added servergroupautodelperm function
+ added -seconds_empty parameter to channelList documentation
+ added missing parameters to channelInfo documentation
+ added missing parameters to clientInfo documentation
+ added missing parameters to serverInfo documentation
+ added alias functions starting with 'priviledge' for token functions
~ changed internal code of permOverview (you don't need to change your code)
~ made time parameter in banAddByIp function optional
~ made time parameter in banAddByUid function optional
~ made time parameter in banAddByName function optional
~ made time parameter in banClient function optional


Version: 0.9.0.0 - Tue, 19 Jul 2016

+ added "-mapping" to the serversnapshotdeploy command
+ added customSearch function (thanks to VJe@n, who let me know that this function is missing)
+ added customInfo function (thanks to VJe@n, who let me know that this function is missing)
~ changed clientList example, because b_virtualserver_client_list's new default value
~ did some error prevention in tokenAdd function
! fixed bug in tokenAdd function that just the first ident of customset will be imported


Version: 0.8.0.0 - Thu, 22 Oct 2015

* switched to a new documentation engine: doxygen
+ added clientAvatar function
+ added example for clientList with avatar
~ changed ftGetFileInfo, the function will now just accept one file instead of unlimited
~ fixed bug in serverCreate (thanks to Michael Nicolai)
~ fixed bug in logView (thanks to Thomas from zenserv.fr)


Version: 0.7.0.0 - Tue, 28 Jan 2014

~ fixed missing parameter in permission array in clientAddPerm
~ fixed bug in channelClientDelPerm
~ fixed bug in channelGroupDelPerm
~ updated clientList documentation
~ updated channelInfo documentation


Version: 0.6.9.0 - Sat, 17 Aug 2013

~ changed documentation of sendMessage
~ fixed conversion error in tokenAdd function (thanks to PickleKillz from Iron Vengeance)
~ fixed misnamed parameter in channelGroupAddPerm function (thanks to Xerxates)


Version: 0.6.8.1 - Sun, 03 Mar 2013

~ hotfix, fixed a bug in permOverview that permid can be used with permid = 0


Version: 0.6.8.0 - Sun, 03 Mar 2013

! permissionList with parameter new can just be used in version >= 3.0.7
+ added parameter new to permissionList
~ changed logic of permissionList to be compatible with server version 3.0.7
~ fixed a bug in selectServer (thanks to Psychokiller)

examples/clientkick.php
<?php
/**
  * clientkick.php
  *
  * Is a small script to demonstrate how to kick a client via ts3admin.class
  *
  * by par0noid solutions - ts3admin.info
  *
*/

/*-------SETTINGS-------*/
$ts3_ip = '127.0.0.1';
$ts3_queryport = 10011;
$ts3_user = 'serveradmin';
$ts3_pass = 'password';
$ts3_port = 9987;
/*----------------------*/

#Include ts3admin composer library
require_once __DIR__ . '/../vendor/autoload.php';
use par0noid\ts3admin\ts3admin;

#build a new ts3admin object
$tsAdmin = new ts3admin($ts3_ip, $ts3_queryport);

if($tsAdmin->getElement('success', $tsAdmin->connect())) {
	#login as serveradmin
	$tsAdmin->login($ts3_user, $ts3_pass);
	
	#select teamspeakserver
	$tsAdmin->selectServer($ts3_port);
	
	#check if id parameter is set
	if(isset($_GET['id']) and !empty($_GET['id']) and ctype_digit($_GET['id'])) {
		#kick player from server
		if(!$tsAdmin->clientKick($_GET['id'], 'server')) {
			#show error if client kick was failed
			echo '<script>alert(\'Client not found\')</script>';
		}
	}
	
	#get clientlist
	$clients = $tsAdmin->clientList();
	
	#print client count
	echo count($clients['data']) . ' clients on selected server<br><br>';
	
	#print clients to browser
	foreach($clients['data'] as $client) {
		echo '<a href="clientkick.php?id='.$client['clid'].'">'.$client['client_nickname'].'</a><br>';
	}
}else{
	echo 'Connection could not be established.';
}

/**
 * This code retuns all errors from the debugLog
 */
if(count($tsAdmin->getDebugLog()) > 0) {
	foreach($tsAdmin->getDebugLog() as $logEntry) {
		echo '<script>alert("'.$logEntry.'");</script>';
	}
}
?>